app/scripts/controllers/clients.js

```diff
@@ -55,21 +55,327 @@ angular.module('openhimConsoleApp')
+    /*----------------------------------Roles----------------------------*/
+    $scope.addNewRole = false;
+    $scope.newRoles = [];
+    $scope.newRolesIndex = 0;
+    
+    var apiCall = function(method, parameters, body, callback) {
+      var success = function() {
+        if (callback) {
+          callback();
+        }
+      };
+      
+      var error = function(error) {
+        Alerting.AlertReset('server');
+        Alerting.AlertAddMsg('server', 'error', error);
+      };
+      
+      switch(method) {
+        case 'update':
+          Api.Roles.update(parameters, body, success, error);
+          break;
+        case 'save':
+          Api.Roles.save(parameters, body, success, error);
+          break;
+        case 'remove':
+          Api.Roles.remove(parameters, body, success, error);
+          break;
+      }
+    };
+    
+  
+    /* -------------------------Load Channels---------------------------- */
+    var channelQuerySuccess = function(channels){
+      $scope.channels = channels;
+      if( channels.length === 0 ){
+        Alerting.AlertAddMsg('bottom', 'warning', 'There are currently no channels created');
+      } else {
+        Alerting.AlertReset('bottom');
+      }
+    };
+
+    Api.Channels.query(channelQuerySuccess, queryError); // request channels for table columns
+    /* -------------------------End Load Channels---------------------------- */
+      
+    
+    /* -------------------------Load Roles---------------------------- */  
+    $scope.$watch('channels', function (newVal) {
+      if(newVal) {
+        loadRoles();  // channels need to be loaded before roles to set up the table columns
+      }
+    });
+    
+    var loadRoles = function () {
+      var roleQuerySuccess = function(roles) {
+        $scope.roles = roles;
+        if( roles.length === 0 ) {
+          Alerting.AlertAddMsg('bottom', 'warning', 'There are currently no roles created');
+        } else {
+          Alerting.AlertReset('bottom');
+        }
+      };
 
+      Api.Roles.query(roleQuerySuccess, queryError); // request roles
 
+      $scope.$on('rolesChanged', function() {
+        Api.Roles.query(roleQuerySuccess, queryError);  // listen for changed roles and reload roles
+      });
+    };
+    /* -------------------------End Load Roles---------------------------- */
+    
+    
+    /* -------------------------Assign Clients to Roles---------------------------- */
+    $scope.$watch('clients', function (newVal) {
+      if(newVal) {
+        waitForRoles(); // wait for roles before assigning clients
+      }
+    });
+    
+    var waitForRoles = function() {
+      $scope.$watch('roles', function (newVal) {
+        if(newVal) {
+          buildClientsRolesObject();
+          buildChannelsRolesObject();
+          angular.forEach($scope.roles, function(role) {
+            role.displayName = role.name;
+          });
+        }
+      });
+    };
+    
+    var buildClientsRolesObject = function() {
```

Comment:
It's better practise to define a function above where it is used in code. Here, it works fine because the function is only called when $watch fires, however, perhaps we should still move the code for stylistic reasons? (This is done above as well).